src/java.base/share/classes/jdk/internal/lang/stable/StableValueImpl.java

```diff
@@ -0,0 +1,419 @@
+        if (isSet() || isError()) {
+           return orThrow();
+        }
+        final var m = acquireMutex();
+        if (isMutexNotNeeded(m)) {
+            return orThrow();
+        }
+        synchronized (m) {
+            if (isSet() || isError()) {
+                return orThrow();
+            }
+            setValue(value);
+            return value;
+        }
+    }
+
+    @ForceInline
+    @Override
+    public boolean trySet(V value) {
+        if (isSet() || isError()) {
+            return false;
+        }
+        final var m = acquireMutex();
+        if (isMutexNotNeeded(m)) {
+            return false;
+        }
+        synchronized (m) {
+            if (isSet() || isError()) {
+                return false;
+            }
+            setValue(value);
+            return true;
+        }
+    }
+
+    @ForceInline
+    @Override
+    public V computeIfUnset(Supplier<? extends V> supplier) {
+        // Optimistically try plain semantics first
+        final V v = value;
+        if (v != null) {
+            // If we happen to see a non-null value under
+            // plain semantics, we know a value is set.
+            return v;
+        }
+        if (state == NULL) {
+            // If we happen to see a state value of NULL under
+            // plain semantics, we know a value is set to `null`.
+            return null;
+        }
+        // Now, fall back to volatile semantics.
+        return computeIfUnsetVolatile(supplier);
+    }
+
+    @DontInline // Slow-path taken at most once per thread if set
+    private V computeIfUnsetVolatile(Supplier<? extends V> supplier) {
+        // This is intentionally an old switch statement as it generates
+        // more compact byte code.
+        switch (stateVolatile()) {
+            case UNSET:    { return computeIfUnsetVolatile0(supplier); }
+            case NON_NULL: { return valueVolatile(); }
+            case NULL:     { return null; }
+            case ERROR:    { throw StableUtil.error(this); }
+        }
+        throw shouldNotReachHere();
+    }
+
+    private V computeIfUnsetVolatile0(Supplier<? extends V> supplier) {
+        final var m = acquireMutex();
+        if (isMutexNotNeeded(m)) {
+            return orThrow();
+        }
+        synchronized (m) {
+            // A value is already set
+            if (state != UNSET) {
+                return orThrow();
+            }
+            // A value is not set
+            if (supplying) {
+                throw stackOverflow(supplier, null);
+            }
+            try {
+                supplying = true;
+                try {
+                    V newValue = supplier.get();
+                    setValue(newValue);
+                    return newValue;
+                } catch (Throwable t) {
+                    putState(ERROR);
+                    putMutex(t.getClass());
+                    throw t;
+                }
+            } finally {
+                supplying = false;
```

Comment:
Resetting a stable field is a bad idea. I recommend renaming this to `supplierCalled` or `supplied` so we never transition this false -> true

Reply:
Yes, according to the `@Stable` annotation’s JavaDoc, this is UB:
https://github.com/openjdk/jdk/blob/8a4315f833f3700075d65fae6bc566011c837c07/src/java.base/share/classes/jdk/internal/vm/annotation/Stable.java#L74-L80

Reply:
Fyi what usually happens is that if a stable field or similarly constant-folded field is promoted to constant, the constant promotion can happen to any of the previous valid values written.

MethodHandle optimisitically sets a trusted final field this way:
https://github.com/openjdk/jdk/blob/8a4315f833f3700075d65fae6bc566011c837c07/src/java.base/share/classes/java/lang/invoke/MethodHandle.java#L1868-L1870

Also a similar example in user code targeting older Java releases, before JDK 16's strong encapsulation so that enums could have been added by reflection:
https://github.com/MinecraftForge/MinecraftForge/issues/3885#issuecomment-355602542

Reply:
Somehow the `@Stable` annotation sneaked into the `supplying` field. But actually keeping it that way and just set the value once would be better.